lib/logutils/ratelimitedlogger.go

```diff
@@ -0,0 +1,308 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package logutils
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	fieldLogSkipped = "logs-skipped"
+	fieldLogNextLog = "next-log"
+	defaultInterval = 5 * time.Minute
+)
+
+// RateLimitedLoggerConfig defines the
+type RateLimitedLoggerConfig struct {
+	// Defaults to 5 minutes if not specified.
+	Interval time.Duration
+
+	// Defaults to the standard logger if not specified.
+	Logger *logrus.Logger
+}
+
+// NewRateLimitedLogger returns a RateLimitedLogger which can be used for rate limited logging.
+//
+// Methods are essentially the same as the logrus logging methods, but there is no Panic or Fatal log since these don't
+// make much sense for rate limited logging.
+//
+// Log requests are only processed if allowed by the logging level. The first processed log will always be written.
+// Subsequent processed logs will not be written within the configured time interval. Once the time interval has passed
+// the next log will be written. The logs include additional fields specifying the number of skipped logs and the
+// minimum time for the next expected log.  The Force() method can be used to ensure the log is written - this resets
+// the time for the next log.
+//
+// Typical use might be as follows:
+//
+//   Logger := NewRateLimitedLogger(nil).WithField("key": "my-key")
+//   for {
+//     Logger.Infof("Checking some stuff: %s", myStuff)
+//     complete = doSomeStuff()
+//     if complete {
+//       break
+//     }
+//   }
+//
+//   // Use force to ensure our final log is printed and it contains the summary info about the number of skipped logs.
+//   Logger.Force().Info("Finished checking stuff")
+//
+// The config is an optional parameter. If not specified, default values are used (see RateLimitedLoggerConfig for
+// details about the default values).
+func NewRateLimitedLogger(config *RateLimitedLoggerConfig) *RateLimitedLogger {
```

Comment:
Suggest using the options pattern to avoid needing to pass nil:
```
type RateLimitedLoggerOpt func(*RateLimitedLogger)

func OptInterval(d time.Duration) RateLimitedLoggerOpt {
  return func(r *RateLimitedLogger) {
    r.data.interval = d
  }
}

func NewRateLimitedLogger(opts ...RateLimitedLoggerOpt) *RateLimitedLogger 
...

r := &RateLimitedLogger{
		data: &intervalData{
			nextLog:  time.Now(),
			interval: defaultInterval,
		},
		entry: logrus.NewEntry(logger),
	}
	for _, opt :=range opts {
		opt(r)
	}	
	return r
}
```